### PR TITLE
❇️ Add Function to Save Engines Output in the `zarr` Format

### PR DESCRIPTION
**Pull request by AbishekRajVG**
Updates to the `tiatoolbox.utils.misc.py`
- Added functions to save the Tiatoolbox Engine outputs to Zarr and AnnotationStore files.

### CODE REVIEW

File: tiatoolbox/utils/misc.py [hunk 1]
```diff
@@ -1212,9 +1216,16 @@ def patch_pred_store(
             conversion of annotations saved at non-baseline resolution to baseline.
             Should be model_mpp/slide_mpp.
         class_dict (dict): Optional dictionary mapping class indices to class names.
+        save_dir (str or pathlib.Path): Optional Output directory to save the Annotation
+            Store results. if the save_dir is not provided, then an SQLiteStore object
+            containing Annotations for each patch is returned.
+        output_file (str): Optional file name to save the Annotation Store results.
+            if the output_file is not provided, then an SQLiteStore object
+            containing Annotations for each patch is returned.
```
Comment (measty) [inline tiatoolbox/utils/misc.py:1225]: is there any particular reason to have both save_dir and output_file? why not just save_path, or something similar?

File: tiatoolbox/utils/misc.py [hunk 2]
```diff
@@ -1257,4 +1268,58 @@ def patch_pred_store(
...
+    """Persists the patch predictor output to a zarr file.
+
+    Args:
+        raw_predictions (dict): A dictionary of patch prediction information.
+        save_dir (str or pathlib.Path): Output directory to save the Annotation
+            Store results. if the save_dir is not provided, then an SQLiteStore object
+            containing Annotations for each patch is returned.
+        output_file (str): File name to save the Annotation Store results.
+            if the output_file is not provided, then an SQLiteStore object
+            containing Annotations for each patch is returned.
```
Comment (measty) [inline tiatoolbox/utils/misc.py:1297]: docstring needs updating, its not saving/returning SQLiteStores etc

**Review by measty — COMMENTED** [review]
just left a couple of comments

File: tiatoolbox/utils/misc.py (hunk 1 above)
Comment (shaneahmed) [reply tiatoolbox/utils/misc.py:1225]: I agree for a misc function we can use just one variable save_path, from engines we can pass `save_path / output_name`

File: tiatoolbox/utils/misc.py [hunk 3]
```diff
@@ -1257,4 +1268,58 @@ def patch_pred_store(
     store = SQLiteStore()
     keys = store.append_many(annotations, [str(i) for i in range(len(annotations))])
 
+    # if a save director is provided, then dump store into a file
+    if save_dir and output_file:
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:1272]: This if block needs to be covered with unit tests.

File: tiatoolbox/utils/misc.py [hunk 4]
```diff
@@ -1257,4 +1268,58 @@ def patch_pred_store(
...
+        save_dir.mkdir(parents=True, exist_ok=True)
+        store.dump(path_to_output_file)
+        return path_to_output_file
+
     return store
+
+
+def patch_pred_store_zarr(
+    raw_predictions: dict,
+    save_dir: Path,
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:1284]: same as above, please replace this with `save_path`, we can pass `save_dir / output_name` from engines.

File: tiatoolbox/utils/misc.py [hunk 5]
```diff
@@ -1257,4 +1268,58 @@ def patch_pred_store(
...
+    output_file: str,
+    **kwargs: dict,
+) -> Path:
+    """Persists the patch predictor output to a zarr file.
+
+    Args:
+        raw_predictions (dict): A dictionary of patch prediction information.
+        save_dir (str or pathlib.Path): Output directory to save the Annotation
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:1292]:
```suggestion
        save_dir (str or Path): Output directory to save the Annotation
```

File: tiatoolbox/utils/misc.py [hunk 6]
```diff
@@ -1212,9 +1216,16 @@ def patch_pred_store(
             conversion of annotations saved at non-baseline resolution to baseline.
             Should be model_mpp/slide_mpp.
         class_dict (dict): Optional dictionary mapping class indices to class names.
+        save_dir (str or pathlib.Path): Optional Output directory to save the Annotation
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:1219]:
```suggestion
        save_dir (str or Path): Optional Output directory to save the Annotation
```
It's imported as `Path`

**Comment by AbishekRajVG** [comment]
Thanks for your comments @shaneahmed and @measty. I've now updated the code accordingly and pushed my changes.

File: tests/test_utils.py [hunk 7]
```diff
@@ -1711,3 +1711,74 @@ def test_patch_pred_store_sf() -> None:
...
+    save_path = tmp_path / "patch_output" / "output.zarr"
+
+    store_path = misc.patch_pred_store_zarr(patch_output, save_path=save_path)
+
+    print("Zarr path: ", store_path)
+    assert Path.exists(store_path), "Zarr output file does not exist"
+
+
+def test_patch_pred_store_zarr_ext(tmp_path: pytest.TempPathFactory) -> None:
+    """Test patch_pred_store and persists predicted output and ensures the output file
```
Comment (shaneahmed) [inline tests/test_utils.py:1735]: First line is always a short line describing the function.
Next line is empty 
Description is after that.

"""Summary line.

Description.

"""

**Comment by shaneahmed** [comment]
> Thanks for your comments @shaneahmed and @measty. I've now updated the code accordingly and pushed my changes.

@AbishekRajVG there is an error for docstring. Please can you address this?

File: tests/test_utils.py (hunk 7 above)
Comment (shaneahmed) [reply tests/test_utils.py:1735]: Please see this page for further examples: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

**Comment by AbishekRajVG** [comment]
> > Thanks for your comments @shaneahmed and @measty. I've now updated the code accordingly and pushed my changes.
> 
> @AbishekRajVG there is an error for docstring. Please can you address this? you can use `ruff check .` to test locally before pushing to Git.

Sorry about that, yes ruff check cleared and updates are pushed.

File: tiatoolbox/utils/misc.py [hunk 8]
```diff
@@ -1257,4 +1263,55 @@ def patch_pred_store(
     store = SQLiteStore()
     keys = store.append_many(annotations, [str(i) for i in range(len(annotations))])
 
+    # if a save director is provided, then dump store into a file
+    if save_path:
+        # ensure parent directory exisits
+        save_path.parent.absolute().mkdir(parents=True, exist_ok=True)
+        # ensure proper db extension
+        if save_path.suffix != ".db":
+            save_path = save_path.parent.absolute() / (save_path.stem + ".db")
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:1272]: This line needs to be covered in tests. The rest of this PR is fine.

File: tests/test_utils.py [hunk 9]
```diff
@@ -1704,10 +1704,113 @@ def test_patch_pred_store_sf() -> None:
...
 
     # Check that its an SQLiteStore containing the expected annotations
     assert isinstance(store, SQLiteStore)
     assert len(store) == 3
     for annotation in store.values():
         assert annotation.geometry.area == 4
+
+
+def test_patch_pred_store_zarr(tmp_path: pytest.TempPathFactory) -> None:
+    """Test patch_pred_store_zarr."""
```
Comment (shaneahmed) [inline tests/test_utils.py:1717]: Please update docstring.

File: tiatoolbox/utils/misc.py [hunk 10]
```diff
@@ -1197,24 +1199,34 @@ def add_from_dat(
...
 
-def patch_pred_store(
+def dict_to_store(
     patch_output: dict,
     scale_factor: tuple[int, int],
     class_dict: dict | None = None,
-) -> AnnotationStore:
+    save_path: Path | None = None,
+) -> AnnotationStore | Path:
     """Create an SQLiteStore containing Annotations for each patch.
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:1209]: Please add in the description that this function takes the output of TIAToolbox engines and converts these to AnnotationStore.

File: tiatoolbox/utils/misc.py [hunk 11]
```diff
@@ -1257,4 +1269,56 @@ def patch_pred_store(
...
+
     return store
+
+
+def dict_to_zarr(
+    raw_predictions: dict,
+    save_path: Path,
+    **kwargs: dict,
+) -> Path:
+    """Persists the patch predictor output to a zarr file.
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:1289]:
```suggestion
    """Saves the output of TIAToolbox engines to a zarr file.
```

File: tiatoolbox/utils/misc.py [hunk 12]
```diff
@@ -1197,24 +1199,34 @@ def add_from_dat(
...
-) -> AnnotationStore:
-    """Create an SQLiteStore containing Annotations for each patch.
+    save_path: Path | None = None,
+) -> AnnotationStore | Path:
+    """Converts (and optionally saves) output of TIAToolbox engines as AnnotationStore.
 
     Args:
-        patch_output (dict): A dictionary of patch prediction information. Important
+        patch_output (dict):
+            A dictionary of patch prediction information. Important
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:1212]:
```suggestion
            A dictionary from patch prediction output. Important
```

File: tiatoolbox/utils/misc.py [hunk 13]
```diff
@@ -1257,4 +1269,56 @@ def patch_pred_store(
...
+def dict_to_zarr(
+    raw_predictions: dict,
+    save_path: Path,
+    **kwargs: dict,
+) -> Path:
+    """Saves the output of TIAToolbox engines to a zarr file.
+
+    Args:
+        raw_predictions (dict):
+            A dictionary of output from TIAToolbox Engines.
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:1293]:
```suggestion
            A dictionary in the TIAToolbox Engines output format.
```

**Pull request merged**